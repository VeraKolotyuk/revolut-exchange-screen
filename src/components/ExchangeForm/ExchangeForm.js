import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';

import AmountInput from '../AmountInput/AmountInput';
import CurrencySelect from '../CurrencySelect/CurrencySelect';
import UserBalance from '../UserBalance/UserBalance';
import {onCurrencySelectChange, onInputChange} from '../../actions/formActions';
import {getUserBalance} from '../../actions/userActions';
import {exchange, fetchRates} from '../../actions/exchangeActions';
import './style.css';
import {formatValue, getCurrentBalance, checkBalanceExceeds} from '../../utils/exchangeUtils';
import {INPUT_TO_VALUE_TYPE,
        INPUT_FROM_VALUE_TYPE,
        CURRENCY_FROM_VALUE_TYPE,
        CURRENCY_TO_VALUE_TYPE
} from '../../utils/constants';

const ExchangeForm = ({wallet,
                      onInputChange,
                      currencyRate,
                      inputFromValue,
                      inputToValue,
                      currencyFromValue,
                      currencyToValue,
                      onSelectChange
}) => {
    const inputToChangeHandler = (value) => {
        onInputChange(value, INPUT_TO_VALUE_TYPE);
        let convertedValue = value / currencyRate * -1;
        onInputChange(formatValue(convertedValue), INPUT_FROM_VALUE_TYPE);
    };

    const inputFromChangeHandler = (value) => {
        onInputChange(value, INPUT_FROM_VALUE_TYPE);
        let convertedValue = currencyRate * value * -1;
        onInputChange(formatValue(convertedValue), INPUT_TO_VALUE_TYPE);
    };

    return (
        <div className='exchange-form'>
            <div className='exchange-form-row'>
                <div>
                    <CurrencySelect
                        value={currencyFromValue}
                        onSelectChangeHandler={(selected) => {
                            onSelectChange(selected.value, CURRENCY_FROM_VALUE_TYPE);
                        }}
                    />
                    <UserBalance currency={currencyFromValue} balance={getCurrentBalance(wallet, currencyFromValue)} />
                </div>
                <AmountInput
                    value={inputFromValue}
                    onInputChangeHandler = {(e) => {
                        inputFromChangeHandler(e.target.value);
                    }}
                    showError={checkBalanceExceeds(inputFromValue, wallet, currencyFromValue)}
                />
            </div>
            <div className='exchange-form-row'>
                <div>
                    <CurrencySelect
                        value={currencyToValue}
                        onSelectChangeHandler={(selected) => {
                            onSelectChange(selected.value, CURRENCY_TO_VALUE_TYPE);
                        }}
                    />
                    <UserBalance currency={currencyToValue}
                                 balance={getCurrentBalance(wallet, currencyToValue)}
                    />
                </div>
                <AmountInput
                    value={inputToValue}
                    onInputChangeHandler = {(e) => {
                        inputToChangeHandler(e.target.value);
                    }}
                    showError={checkBalanceExceeds(inputToValue, wallet, currencyToValue)}
                />
            </div>
        </div>
    );
};

const mapDispatchToProps = dispatch => ({
    onInputChange: bindActionCreators(onInputChange, dispatch),
    onSelectChange: bindActionCreators(onCurrencySelectChange, dispatch),
    getUserBalance: bindActionCreators(getUserBalance, dispatch),
    fetchRates: bindActionCreators(fetchRates, dispatch),
    exchange: bindActionCreators(exchange, dispatch),
});

const mapStateToProps = store => ({
    inputFromValue: store.form.inputFromValue,
    inputToValue: store.form.inputToValue,
    currencyFromValue: store.form.currencyFromValue,
    currencyToValue: store.form.currencyToValue,
    wallet: store.user.wallet,
    rates: store.exchange.rates
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ExchangeForm);
