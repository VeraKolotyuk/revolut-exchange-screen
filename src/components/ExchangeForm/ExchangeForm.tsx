import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';

import AmountInput from '../AmountInput/AmountInput';
import CurrencySelect from '../CurrencySelect/CurrencySelect';
import UserBalance from '../UserBalance/UserBalance';
import {onCurrencySelectChange, onInputChange} from '../../actions/formActions';
import {getUserBalance} from '../../actions/userActions';
import {exchange, fetchRates} from '../../actions/exchangeActions';
import './style.css';
import {formatValue, getCurrentBalance, checkBalanceExceeds} from '../../utils/exchangeUtils';
import {INPUT_TO_VALUE_TYPE,
        INPUT_FROM_VALUE_TYPE,
        CURRENCY_FROM_VALUE_TYPE,
        CURRENCY_TO_VALUE_TYPE
} from '../../utils/constants';

type Props = {
    wallet: [any],
    onInputChange: (a: string, b: string) => void,
    currencyRate: number | null,
    inputFromValue: number,
    inputToValue: number,
    currencyFromValue: string,
    currencyToValue: string,
    onSelectChange: (a: string, b: string) => void,
};

interface Options {
    value: string,
    label: string
}

const ExchangeForm = ({wallet,
                      onInputChange,
                      currencyRate,
                      inputFromValue,
                      inputToValue,
                      currencyFromValue,
                      currencyToValue,
                      onSelectChange
}: Props) => {
    const inputToChangeHandler = (value: string) => {
        onInputChange(value, INPUT_TO_VALUE_TYPE);
        //TODO:: extract method
        let convertedValue = Number(value) / Number(currencyRate) * -1;
        onInputChange(formatValue(convertedValue), INPUT_FROM_VALUE_TYPE);
    };

    const inputFromChangeHandler = (value: string) => {
        onInputChange(value, INPUT_FROM_VALUE_TYPE);
        let convertedValue = Number(currencyRate) * Number(value) * -1;
        onInputChange(formatValue(convertedValue), INPUT_TO_VALUE_TYPE);
    };

    return (
        <div className='exchange-form'>
            <div className='exchange-form-row'>
                <div>
                    <CurrencySelect
                        value={currencyFromValue}
                        onSelectChangeHandler={(selected: Options) => {
                            onSelectChange(selected.value, CURRENCY_FROM_VALUE_TYPE);
                        }}
                        valueToRemove={currencyToValue}
                    />
                    <UserBalance currency={currencyFromValue} balance={getCurrentBalance(wallet, currencyFromValue)} />
                </div>
                <AmountInput
                    value={inputFromValue}
                    onInputChangeHandler = {(e: { target: HTMLInputElement }) => {
                        inputFromChangeHandler(e.target.value);
                    }}
                    showError={checkBalanceExceeds(inputFromValue, wallet, currencyFromValue)}
                />
            </div>
            <div className='exchange-form-row'>
                <div>
                    <CurrencySelect
                        value={currencyToValue}
                        onSelectChangeHandler={(selected: Options) => {
                            onSelectChange(selected.value, CURRENCY_TO_VALUE_TYPE);
                        }}
                        valueToRemove={currencyFromValue}
                    />
                    <UserBalance currency={currencyToValue}
                                 balance={getCurrentBalance(wallet, currencyToValue)}
                    />
                </div>
                <AmountInput
                    value={inputToValue}
                    onInputChangeHandler = {(e: { target: HTMLInputElement }) => {
                        inputToChangeHandler(e.target.value);
                    }}
                    showError={checkBalanceExceeds(inputToValue, wallet, currencyToValue)}
                />
            </div>
        </div>
    );
};

const mapDispatchToProps = dispatch => ({
    onInputChange: bindActionCreators(onInputChange, dispatch),
    onSelectChange: bindActionCreators(onCurrencySelectChange, dispatch),
    getUserBalance: bindActionCreators(getUserBalance, dispatch),
    fetchRates: bindActionCreators(fetchRates, dispatch),
    exchange: bindActionCreators(exchange, dispatch),
});

const mapStateToProps = store => ({
    inputFromValue: store.form.inputFromValue,
    inputToValue: store.form.inputToValue,
    currencyFromValue: store.form.currencyFromValue,
    currencyToValue: store.form.currencyToValue,
    wallet: store.user.wallet,
    rates: store.exchange.rates
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ExchangeForm);
