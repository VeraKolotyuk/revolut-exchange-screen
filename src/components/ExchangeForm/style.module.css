.exchange-form {
    margin-top: 2.5rem;
}

.exchange-form .exchange-form-row {
    display: flex;
    justify-content: space-between;
    width: 25rem;
    background: white;
    margin-bottom: 2rem;
    border-radius: 1rem;
    padding: 1.5rem;
}