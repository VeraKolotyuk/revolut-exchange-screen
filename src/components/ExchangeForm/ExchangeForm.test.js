import ExchangeForm from './ExchangeForm';
import { shallow, configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

const mockStore = configureMockStore();

configure({adapter: new Adapter()});

let wrapper;

beforeEach(() => {
    wrapper = shallow(<Provider store={mockStore()}><ExchangeForm/></Provider>);
});

describe('ExchangeForm rendering', () => {
    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
});