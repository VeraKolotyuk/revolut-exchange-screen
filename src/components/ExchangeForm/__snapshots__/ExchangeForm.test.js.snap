// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeForm rendering renders correctly 1`] = `ShallowWrapper {}`;
